packages/js/src/utils/create-async-iterable/create-async-iteratable.ts

```diff
@@ -0,0 +1,67 @@
+export interface AsyncPushCallbacks<T> {
+  next: (value: T) => void;
+  done: () => void;
+  error: (err: unknown) => void;
+}
+
+export function createAsyncIterable<T = unknown>(
+  listener: (ls: AsyncPushCallbacks<T>) => void
+): AsyncIterable<T> {
+  let done = false;
+  let error: unknown | null = null;
+
+  const pushQueue: T[] = [];
+  const pullQueue: Array<
+    [
+      (x: { value: T | undefined; done: boolean }) => void,
+      (err: unknown) => void
+    ]
+  > = [];
+
+  return {
+    [Symbol.asyncIterator]() {
+      listener({
+        next: (value) => {
+          if (done || error) return;
+          if (pullQueue.length > 0) {
+            pullQueue.shift()?.[0]({ value, done: false });
+          } else {
+            pushQueue.push(value);
+          }
+        },
+        error: (err) => {
+          if (done || error) return;
+          if (pullQueue.length > 0) {
+            pullQueue.shift()?.[1](err);
+          } else {
+            error = err;
+          }
+        },
+        done: () => {
+          if (pullQueue.length > 0) {
+            pullQueue.shift()?.[0]({ value: undefined, done: true });
+          } else {
+            done = true;
+          }
+        },
+      });
+
+      return {
+        next() {
+          return new Promise<{ value: T | undefined; done: boolean }>(
+            (resolve, reject) => {
+              if (done) resolve({ value: undefined, done: true });
```

Comment:
I think the pushQueue has to be processed first. push push done, should return twice before returning done: true, right?

Reply:
If the `done()` callback is called first, then subsequent `next()` calls should not do anything right?

This is the push-push-done case.
```
next(a); // { value: a, done: false }
next(b); // { value: b, done: false }
done();  // { value: undefined, done: true }
```

And here is the push-done-push case.
```
next(a); // { value: a, done: false }
done();  // { value: undefined, done: true }
next(b); // { value: undefined, done: true }
```

Reply:
I think so, but I think that is handled correctly already. `next` just returns when `done` is true. That's the listener/writer side. My comment is about the reader side. 

I think the reader has to return all the pushed values. If the writer pushes 1,2,3 and then calls "done". The reader should receive:

```
{value: 1, done: false}
{value: 2, done: false}
{value: 3, done: false}
{value: undefined, done: true}
```

The way it's written now, if next(3) and done() will happen right after (in the same microtask) another (which they often will), the reader will never receive 3. 

Right?

Reply:
I can confirm that @vsavkin concern is correct. `pushQueue` needs to be resolved, if there are still items left, before `done` or `error`. I adjusted the `next()` and added a test case.